Various enhancements to the Gradio Ray Serve tutorial (#50276)
ray-project/ray: doc/source/serve/tutorials/gradio-integration.md

# Scale a Gradio App with Ray Serve

This guide shows how to scale up your [Gradio](https://gradio.app/) application using Ray Serve. Keep the internal architecture of your Gradio app intact, with no code changes. Simply wrap the app within Ray Serve as a deployment and scale it to access more resources.

This tutorial uses Gradio apps that run text summarization and generation models and use [Hugging Face's Pipelines](https://huggingface.co/docs/transformers/main_classes/pipelines) to access these models.

:::{note} 
Remember you can substitute this app with your own Gradio app if you want to try scaling up your own Gradio app.
:::

## Dependencies

To follow this tutorial, you need Ray Serve, Gradio, and transformers. If you haven't already, install them by running:
```console
$ pip install "ray[serve]" gradio==3.50.2 torch transformers
```

## Example 1: Scaling up your Gradio app with `GradioServer`

> The first example summarizes text using the [T5 Small](https://huggingface.co/t5-small) model and uses [Hugging Face's Pipelines](https://huggingface.co/docs/transformers/main_classes/pipelines) to access that model. It demonstrates one easy way to deploy Gradio apps onto Ray Serve: using the simple `GradioServer` wrapper. Later, in example 2, we will show how to use `GradioIngress` for more customized use-cases.

Reviewer: ```suggestion
The first example summarizes text using the [T5 Small](https://huggingface.co/t5-small) model and uses [Hugging Face's Pipelines](https://huggingface.co/docs/transformers/main_classes/pipelines) to access that model. It demonstrates one easy way to deploy Gradio apps onto Ray Serve: using the simple `GradioServer` wrapper. Example 2 shows how to use `GradioIngress` for more customized use-cases.
```

Reviewer: Avoid first person and future tense when possible.

First, create a new Python file named `demo.py`. Second, import `GradioServer` from Ray Serve to deploy your Gradio app later, `gradio`, and `transformers.pipeline` to load text summarization models.

```{literalinclude} ../doc_code/gradio-integration.py
:start-after: __doc_import_begin__
:end-before: __doc_import_end__
```

Then, write a builder function that constructs the Gradio app `io`. 

```{literalinclude} ../doc_code/gradio-integration.py
:start-after: __doc_gradio_app_begin__
:end-before: __doc_gradio_app_end__
```

### Deploying Gradio Server

> In order to deploy your Gradio app onto Ray Serve, you need to wrap your Gradio app in a Serve [deployment](serve-key-concepts-deployment). `GradioServer` acts as that wrapper. It serves your Gradio app remotely on Ray Serve so that it can process and respond to HTTP requests.

Reviewer: ```suggestion
To deploy your Gradio app onto Ray Serve, you need to wrap the Gradio app in a Serve [deployment](serve-key-concepts-deployment). `GradioServer` acts as that wrapper. It serves your Gradio app remotely on Ray Serve so that it can process and respond to HTTP requests.
```

:end-before: __doc_app_end__
```

> Finally, deploy your Gradio Server. Run the following in your terminal (assuming that you saved the file as `demo.py`):

Reviewer: ```suggestion
Finally, deploy your Gradio Server. Run the following in your terminal, assuming that you saved the file as `demo.py`:
```

Reviewer: Avoid excessive use of parentheses.

See the [Production Guide](serve-in-production) for more information on how to deploy your app in production.

## Example 2: Parallelizing models with Ray Serve
You can run multiple models in parallel with Ray Serve by using [model composition](serve-model-composition) in Ray Serve.

Suppose you want to run the following program.

1. Take two text generation models, [`gpt2`](https://huggingface.co/gpt2) and [`distilgpt2`](https://huggingface.co/distilgpt2).
2. Run the two models on the same input text, so that the generated text has a minimum length of 20 and maximum length of 100.
3. Display the outputs of both models using Gradio.

> Let's compare an unparallelized approach using vanilla Gradio to a parallelized approach using Ray Serve.

Reviewer: ```suggestion
The following is a comparison of an unparallelized approach using vanilla Gradio to a parallelized approach using Ray Serve.
```

> Let's compare an unparallelized approach using vanilla Gradio to a parallelized approach using Ray Serve.

Reviewer 2: ```suggestion
Let's compare using vanilla Gradio to a parallelized approach using Ray Serve.
```

> ### Unparallelized approach

Reviewer 2: ```suggestion
### Vanilla Gradio
```
